### ISSUE TITLE
AST representing sigils with `"""` as a delimiter are converted into invalid code by Macro.to_string/1

### ISSUE THREAD

**Issue by tmbb** [elixir-lang/elixir#12553]
### Elixir and Erlang/OTP versions

Erlang/OTP 25 [erts-13.0.4] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit:ns]

Elixir 1.14.1 (compiled with Erlang/OTP 25)

### Operating system

Ubuntu 20.04

### Current behavior

The elixir syntax rules require a newline before the closing `"""` in a heredoc. This means that the following is valid:

```elixir
~"""
abc
"""
```

but this is invalid:

```elixir
~H"""
abc"""
```

To write the following using heredocs, one has to escale the last newline with a backslash:

```elixir
~H"""
abc\
"""
```

When we build valid AST describing a sigil delimited by `"""` (either by building it from scratch or by using `Code.string_to_quoted/1` on some valid elixir code), if the binary inside the sigil doesn't end with a newline, then running `Macro.to_string/1` on the AST will generate invalid elixir code. See the following IEx session for a concrete example:

```elixir
iex(4)> ast = {:sigil_H, [delimiter: ~S(""")], [{:<<>>, [], ["abc"]}, []]}                                  
{:sigil_H, [delimiter: "\"\"\""], [{:<<>>, [], ["abc"]}, []]}
iex(5)> code = Macro.to_string(ast)
"~H\"\"\"\nabc\"\"\""
iex(6)> IO.puts(code)
~H"""
abc"""
:ok
iex(7)> Code.string_to_quoted(code)
{:error,
 {[line: 2, column: 7],
  "missing terminator: \"\"\" (for heredoc starting at line 1)", ""}}
```

This is a bug because the AST is valid and `Macro.to_string/1` should never generate invalid code from valid AST.

### Expected behavior

I'd expect the following behaviour:

```elixir
iex(4)> ast = {:sigil_H, [delimiter: ~S(""")], [{:<<>>, [], ["abc"]}, []]}                                  
{:sigil_H, [delimiter: "\"\"\""], [{:<<>>, [], ["abc"]}, []]}
iex(5)> code = Macro.to_string(ast)
"~H\"\"\"\nabc\\n\"\"\""
```

However, even that results in strange behaviour. Although it returns the right string, trying to parse that stirng into Elixir code returns what I consider a wrong result:

```elixir
iex(18)> "~H\"\"\"\nabc\\\n\"\"\"" |> Code.string_to_quoted()
{:ok,
 {:sigil_H, [delimiter: "\"\"\"", line: 1],
  [{:<<>>, [indentation: 0, line: 1], ["abc\\\n"]}, []]}}
```

I'd expect the following:

```elixir
iex(18)> "~H\"\"\"\nabc\\\n\"\"\"" |> Code.string_to_quoted()
{:ok,
 {:sigil_H, [delimiter: "\"\"\"", line: 1],
  [{:<<>>, [indentation: 0, line: 1], ["abc"]}, []]}}
```

So I guess some of the parsing logic isn't right either.

**Comment by josevalim** [comment]
It is up to the sigil to decide what happens with escape characters, especially uppercase sigils.

In other words, this is not valid AST: `ast = {:sigil_H, [delimiter: ~S(""")], [{:<<>>, [], ["abc"]}, []]}`. 

Everything else in your issues comes from the assumptions the above is valid but it isn't. If you consider it valid AST, you need to show some code that can generate it. :)

**Comment by tmbb** [comment]
> When we build valid AST describing a sigil delimited by """ (either by building it from scratch or by using Code.string_to_quoted/1 on some valid elixir code)

I was going to swear I had generated that "AST fragment" by calling `Code.string_to_quoted/1` on valid Elixir code, but when I tried to do it again to show you, it turns out the `Code.string_to_quoted/1` does NOT generate that AST. I must have changed it manually at some point. In that case, it's not valid AST as you say, and everything else doesn't make sense.
